packages/effect/src/stream/estream.ts

```diff
@@ -0,0 +1,263 @@
+}
+
+export function once<A>(a: A): EStream<T.NoEnv, T.NoErr, A> {
+  return S.stream.map(S.once(a), a => Ei.right(a));
+}
+
+export function repeatedly<A>(a: A): EStream<T.NoEnv, T.NoErr, A> {
+  return S.stream.map(S.repeatedly(a), a => Ei.right(a));
+}
+
+export function periodically(ms: number): EStream<T.NoEnv, T.NoErr, number> {
+  return S.stream.map(S.periodically(ms), a => Ei.right(a));
+}
+
+export const empty: EStream<T.NoEnv, T.NoErr, never> = S.stream.map(
+  S.never,
+  a => Ei.right(a)
+);
+
+export function raised<E>(e: E): EStream<T.NoEnv, E, never> {
+  return S.once(Ei.left(e));
+}
+
+export function aborted(e: unknown): EStream<T.NoEnv, T.NoErr, never> {
+  return S.stream.map(S.aborted(e), a => Ei.right(a));
+}
+
+export function fromOption<A>(opt: O.Option<A>): EStream<T.NoEnv, T.NoErr, A> {
+  return S.stream.map(S.fromOption(opt), a => Ei.right(a));
+}
+
+export function zipWithIndex<R, E, A>(
+  stream: EStream<R, E, A>
+): EStream<R, E, readonly [A, number]> {
+  return S.stream.map(S.zipWithIndex(stream), a => {
+    if (Ei.isLeft(a[0])) {
+      return Ei.left(a[0].left);
+    } else {
+      return Ei.right([a[0].right, a[1]]);
+    }
+  });
+}
+
+export function concatL<R, E, A, R2, E2>(
+  stream1: EStream<R, E, A>,
+  stream2: Lazy<EStream<R2, E2, A>>
+): EStream<R & R2, E | E2, A> {
+  return S.concatL(stream1, stream2 as any) as any;
+}
+
+export function concat<R, E, A, R2, E2>(
+  stream1: EStream<R, E, A>,
+  stream2: EStream<R2, E2, A>
+): EStream<R & R2, E | E2, A> {
+  return concatL(stream1, () => stream2);
+}
+
+export function repeat<R, E, A>(stream: EStream<R, E, A>): EStream<R, E, A> {
+  return S.repeat(stream);
+}
+
+export function map<R, A, B>(
+  f: FunctionN<[A], B>
+): <E>(stream: EStream<R, E, A>) => EStream<R, E, B> {
+  return stream => map_(stream, f);
+}
+
+export function as<R, E, A, B>(
+  stream: EStream<R, E, A>,
+  b: B
+): EStream<R, E, B> {
+  return map_(stream, _ => b);
+}
+
+export function filter<R, E, A>(
+  stream: EStream<R, E, A>,
+  f: Predicate<A>,
+  propagate: boolean = true
+): EStream<R, E, A> {
+  return S.filter(stream, getEitherP(f, propagate));
+}
+
+export const getEitherP = <E, A>(
+  p: Predicate<A>,
+  propagate: boolean = true
+): Predicate<Ei.Either<E, A>> =>
+  Ei.fold(
+    () => propagate,
+    a => p(a)
+  );
+
+export function filterWith<A>(
+  f: Predicate<A>,
+  propagate: boolean = false
+): <R, E>(stream: EStream<R, E, A>) => EStream<R, E, A> {
+  return stream => filter(stream, f, propagate);
+}
+
+export function filterRefineWith<A, B extends A>(
+  f: Refinement<A, B>,
+  propagate: boolean = false
+): <R, E>(stream: EStream<R, E, A>) => EStream<R, E, B> {
+  return stream => filter(stream, f, propagate) as any;
+}
+
+export const URI = "matechs/EStream";
+export type URI = typeof URI;
+declare module "fp-ts/lib/HKT" {
+  interface URItoKind3<R, E, A> {
+    [URI]: EStream<R, E, A>;
+  }
+}
+
+export const stream: Monad3E<URI> = {
```

Comment:
Shouldn't we directly implement Bifunctor and MonadThrow ?

Reply:
probably yes, I planned to finish porting all the api of stream to here and after finish the instance